core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -0,0 +1,152 @@
+          manifestListsToDelete.add(snapshot.manifestListLocation());
+        }
+      }
+    }
+
+    Set<ManifestFile> candidateManifestFilesForDeletion = readManifests(expiredSnapshots);
+    Set<ManifestFile> manifestFilesAfterExpiration = readManifests(snapshotsAfterExpiration);
+
+    Set<ManifestFile> manifestsToDelete = Sets.newHashSet(candidateManifestFilesForDeletion);
+    for (ManifestFile candidateManifestFile : candidateManifestFilesForDeletion) {
+      if (manifestFilesAfterExpiration.contains(candidateManifestFile)) {
+        manifestsToDelete.remove(candidateManifestFile);
+      }
+    }
+
+    Set<String> dataFilesToDelete =
+        findFilesToDelete(manifestsToDelete, manifestFilesAfterExpiration);
+    deleteFiles(dataFilesToDelete, "Data");
+
+    Set<String> manifestPathsToDelete =
+        manifestsToDelete.stream().map(ManifestFile::path).collect(Collectors.toSet());
+
+    deleteFiles(manifestPathsToDelete, "manifest");
+    deleteFiles(manifestListsToDelete, "manifest list");
+  }
+
+  private Set<ManifestFile> readManifests(Set<Snapshot> snapshots) {
+    Set<ManifestFile> manifestFiles = Sets.newHashSet();
+    for (Snapshot snapshot : snapshots) {
+      try (CloseableIterable<ManifestFile> manifestFilesForSnapshot = readManifestFiles(snapshot)) {
+        for (ManifestFile manifestFile : manifestFilesForSnapshot) {
+          manifestFiles.add(manifestFile.copy());
+        }
+      } catch (IOException e) {
+        throw new RuntimeIOException(
+            e, "Failed to close manifest list: %s", snapshot.manifestListLocation());
+      }
+    }
+
+    return manifestFiles;
+  }
+
+  // Helper to compute files to delete
+  private Set<String> findFilesToDelete(
+      Set<ManifestFile> manifestFilesToDelete, Set<ManifestFile> currentManifestFiles) {
+    Set<String> filesToDelete = ConcurrentHashMap.newKeySet();
+    Tasks.foreach(manifestFilesToDelete)
+        .retry(3)
+        .suppressFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (item, exc) ->
+                LOG.warn("Failed to get deleted files: this may cause orphaned data files", exc))
+        .run(
+            manifest -> {
+              // the manifest has deletes, scan it to find files to delete
+              try (CloseableIterable<String> paths = ManifestFiles.readPaths(manifest, fileIO)) {
+                paths.forEach(filesToDelete::add);
+              } catch (IOException e) {
+                throw new RuntimeIOException(e, "Failed to read manifest file: %s", manifest);
+              }
+            });
+
+    if (filesToDelete.isEmpty()) {
+      return filesToDelete;
+    }
+
+    Tasks.foreach(currentManifestFiles)
+        .retry(3)
+        .suppressFailureWhenFinished()
+        .executeWith(planExecutorService)
+        .onFailure(
+            (item, exc) ->
+                LOG.warn("Failed to get deleted files: this may cause orphaned data files", exc))
```

Comment:
Can this include the path that was not read?

Reply:
Sure,I've updated the log message!